bitcoin/src/taproot/merkle_branch/borrowed.rs

```diff
@@ -0,0 +1,290 @@
+use core::borrow::{Borrow, BorrowMut};
+use internals::slice::SliceExt;
+
+use super::{DecodeError, InvalidMerkleBranchSizeError, InvalidMerkleTreeDepthError, TaprootMerkleBranchBuf, TapNodeHash, TAPROOT_CONTROL_MAX_NODE_COUNT, TAPROOT_CONTROL_NODE_SIZE};
+
+pub use privacy_boundary::TaprootMerkleBranch;
+
+/// Makes sure only the allowed conversions are accessible to external code.
+mod privacy_boundary {
+    use super::*;
+
+    /// The Merkle proof for inclusion of a tree in a Taproot tree hash.
+    #[repr(transparent)]
+    #[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+    pub struct TaprootMerkleBranch([TapNodeHash]);
+
+    impl TaprootMerkleBranch {
+        /// Returns a reference to the slice of hashes.
+        #[inline]
+        pub const fn as_slice(&self) -> &[TapNodeHash] { &self.0 }
+
+        /// Returns a reference to the mutable slice of hashes.
+        #[inline]
+        pub fn as_mut_slice(&mut self) -> &mut [TapNodeHash] { &mut self.0 }
+
+        pub(super) const fn from_hashes_unchecked(hashes: &[TapNodeHash]) -> &Self {
+            unsafe {
+                &*(hashes as *const _ as *const Self)
+            }
+        }
+
+        pub(super) fn from_mut_hashes_unchecked(hashes: &mut [TapNodeHash]) -> &mut Self {
```

Comment:
How is this different from just having a private function outside of the module?

Reply:
Well, true, it doesn't need to be inside since it uses `unsafe` anyway. But I like to having it there anyway because it better shows that this function directly manipulates the internals of the type but also there was a proposal to allow `as` to avoid these `unsafe` blocks and if it ever gets into the language then we will definitely want to use it and have the function right there, so we may as well put it there right now.

(The proposed rule was that if the type is `#[repr(transparent)]` then whichever code sees all its fields as accessible can use `as` to cast the inner slice to the outer type. This is perfectly sensible and what we want, with the implications above.)

Reply:
Fair arguments but none of that convinces me that this is better than not having the module, its unnecessary complexity that is not at all apparent what its there for. IMO simple trumps complex unless there is a good reason for the complexity.

> there was a proposal to allow as to avoid these unsafe blocks and if it ever gets into the language then we will definitely want to use it and have the function right there, so we may as well put it there right now.

I'm not a fan of just-in-case code at the best of times, definitely not for a proposal that is not even in the language yet.

Reply:
Are you talking about module or that specific function? I was talking about that specific function. It strictly doesn't need to be inside the module but it's better to have it there if we have the module anyway. And I want the module because the other function accesses `.0` which is enough reason to have it. If this was the only function that belongs to the module then I would've agreed that adding the module is additional complexity.